### PR TITLE
[Diagnostics] Introduce `getLoc` and `getSourceRange` to failure diagnostic

### PR DESCRIPTION
**Pull request by xedin**
- Some of the diagnostics have special rules about anchor location,
    let's make `getAnchor` virtual and allow sub-class to override
    default implementation.

- Split `emitDiagnostic` into `emitDiagnostic` and `emitDiagnosticAt`,
   where former passes `getLoc` internally.

- Refactor existing diagnostics to use new methods and avoid passing
   location when possible.

### CODE REVIEW

**Comment by xedin** [comment]
This is still a draft because I need to clean up some things here and there but I'd want to get early opinion on the changes. Please take a look!

**Comment by xedin** [comment]
The idea is that `emitDiagnostic` wouldn't have to pass location directly, instead each `FailureDiagnostic` would have to implement `getLoc()` and/or `getSourceRange`. This should make it much easier to refactor locators be anchored on decls/type-reprs/patterns.

File: lib/Sema/CSDiagnostics.cpp [hunk 1]
```diff
@@ -3775,42 +3756,50 @@ bool PartialApplicationFailure::diagnoseAsError() {
...
 
-  emitDiagnostic(anchor->getNameLoc(), diagnostic, kind);
+  emitDiagnosticAt(anchor->getNameLoc(), diagnostic, kind);
   return true;
 }
 
 bool InvalidDynamicInitOnMetatypeFailure::diagnoseAsError() {
-  auto *anchor = getRawAnchor();
-  emitDiagnostic(anchor->getLoc(), diag::dynamic_construct_class,
+  emitDiagnostic(diag::dynamic_construct_class,
```
Comment (hamishknight) [inline lib/Sema/CSDiagnostics.cpp:3764]: Should `InvalidDynamicInitOnMetatypeFailure` implement `getLoc` to use the raw anchor's loc? IMO it makes sense to point to the `.init` if it exists.

File: lib/Sema/CSDiagnostics.cpp [hunk 2]
```diff
@@ -3775,42 +3756,50 @@ bool PartialApplicationFailure::diagnoseAsError() {
...
+  emitDiagnosticAt(Init, diag::note_nonrequired_initializer, Init->isImplicit(),
+                   Init->getFullName());
   return true;
 }
 
 bool InitOnProtocolMetatypeFailure::diagnoseAsError() {
-  auto *anchor = getRawAnchor();
   if (IsStaticallyDerived) {
-    emitDiagnostic(anchor->getLoc(), diag::construct_protocol_by_name,
+    emitDiagnostic(diag::construct_protocol_by_name,
```
Comment (hamishknight) [inline lib/Sema/CSDiagnostics.cpp:3774]: Ditto

File: lib/Sema/CSDiagnostics.cpp [hunk 3]
```diff
@@ -5943,9 +5936,8 @@ bool NonEphemeralConversionFailure::diagnosePointerInit() const {
                     ? diag::cannot_construct_dangling_pointer_warning
                     : diag::cannot_construct_dangling_pointer;
 
-  auto *anchor = getRawAnchor();
-  emitDiagnostic(anchor->getLoc(), diagID, constructedTy, constructorKind)
-      .highlight(anchor->getSourceRange());
+  emitDiagnostic(diagID, constructedTy, constructorKind)
+      .highlight(getSourceRange());
```
Comment (hamishknight) [inline lib/Sema/CSDiagnostics.cpp:5940]: Unfortunately we really do want to use the raw anchor here, as we want to highlight the entire constructor call for this specific diagnostic, as it's mainly talking about the invalid construction of the `UnsafePointer`.

**Review by hamishknight — COMMENTED** [review]
Looks great!

File: lib/Sema/CSDiagnostics.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp:3764]: Makes sense, I'll fix both problems by overriding `getAnchor()` on `InvalidInitRefFailure`!

File: lib/Sema/CSDiagnostics.cpp (hunk 3 above)
Comment (xedin) [reply lib/Sema/CSDiagnostics.cpp]: Fixed!

**Comment by xedin** [comment]
@hamishknight Thanks for the review, I've fixed both problems you have mentioned.

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test Linux platform

**Comment by xedin** [comment]
Since there is no opposition here I'm going to merge this and next up is removal of `Anchor` and `RawAnchor` from `FailureDiagnostic`.

**Pull request merged**